IPython/terminal/shortcuts.py

```diff
@@ -100,12 +100,14 @@ def newline_or_execute(event):
                 b.cancel_completion()
             return
 
-        if not (d.on_last_line or d.cursor_position_row >= d.line_count
-            - d.empty_line_count_at_the_end()):
-            b.newline()
-            return
+        before_text = d.text[:d.cursor_position]
+        status, indent = shell.input_splitter.check_complete(before_text + '\n')
 
-        status, indent = shell.input_splitter.check_complete(d.text + '\n')
+        if not (d.on_last_line or
+                d.cursor_position_row >= d.line_count - d.empty_line_count_at_the_end()
+                ):
+            b.insert_text('\n' + (' ' * (indent or 0)))
+            return
```

Comment:
Scratch that. 

```
In [1]: def foo():
   ...:     pass
   ...: |<enter>
   ...: def bar():
   ...:     pass
```

Above example insert a new line with this patch. 
Execute if I remove above block.